### PR TITLE
[Feature] Adding MR-MTL Personalized Mixin

### PR DESCRIPTION
**Pull request by nerdai**
# PR Type
Feature

# Short Description

## Intro

Now that we have established the mixin/adapter pattern wtih Ditto, in this PR we adopt the same pattern to implement `MrMtlPersonalizedMixin` that works with `FlexibleClient` types.

## Summary

- Adds `MrMtlPersonalizedMixin` and associated `MrMtlPersonalizedProtocol` for typing
- Updates `make_it_personal` class factory method to include `MrMtlPersonalizedMixin`
- Adds `client_dynamic.py`  to `examples/mr_mtl_example` and confirm that it produces exact same result as `client.py` when setting seed — as we did with `examples/ditto_example`

## Testing of dynamically personalized and subclass `MrMtlClient`

image

- left (dynamic) and right (original client inheriting from MrMtlClient)
- examples/mr_mtl_example produces exact same results when setting seed

# Tests Added

- smoke test: `tests/smoke_tests/tests_smoke_tests.py::test_mr_mtl_flexible_mnist`
- unit tests: `tests/mixins/personalized/test_mr_mtl.py`

### CODE REVIEW

File: fl4health/mixins/personalized/mr_mtl.py [hunk 1]
```diff
@@ -0,0 +1,212 @@
+"""MR MTL Personalized Mixin."""
+
+import copy
+import warnings
+from logging import INFO, WARN
```
Comment (emersodb) [inline fl4health/mixins/personalized/mr_mtl.py:5]: We've been using the full WARNING constant just to be consistent. I honestly don't know why there are two...

File: fl4health/mixins/personalized/mr_mtl.py [hunk 2]
```diff
@@ -0,0 +1,212 @@
...
+        # Initialize mixin-specific attributes
+        self.initial_global_model: torch.nn.Module | None = None
+        self.initial_global_tensors: list[torch.Tensor] = []
+
+        # Call parent's init
+        try:
+            super().__init__(*args, **kwargs)
+        except TypeError:
+            # if a parent class doesn't take args/kwargs
+            super().__init__()
```
Comment (emersodb) [inline fl4health/mixins/personalized/mr_mtl.py:59]: I probably asked this before, but don't remember the answer. Under what circumstances do we get here but its okay? I.e. should we just raise the TypeError? Sorry if I have indeed asked this of you previously.

File: fl4health/mixins/personalized/mr_mtl.py [hunk 3]
```diff
@@ -0,0 +1,212 @@
...
+        msg = (
+            f"Class {cls.__name__} inherits from MrMtlPersonalizedMixin but none of its other "
+            f"base classes implement FlexibleClient. This may cause runtime errors."
+        )
+        log(WARN, msg)
+        warnings.warn(msg, RuntimeWarning, stacklevel=2)
+
+    def get_global_model(self: MrMtlPersonalizedProtocol, config: Config) -> nn.Module:
+        """
+        Returns the global model to be used during Ditto training and as a constraint for the local model.
```
Comment (emersodb) [inline fl4health/mixins/personalized/mr_mtl.py:91]: during MR-MTL training? Or perhaps just "Returns the global model on client setup to be used as a constraint for the local model during training"?

Comment (emersodb) [reply fl4health/mixins/personalized/mr_mtl.py:91]: I wonder if we should rename this function to something more suggestive of the fact that we're just setting the initial global model?

File: fl4health/mixins/personalized/mr_mtl.py [hunk 4]
```diff
@@ -0,0 +1,212 @@
...
+    @ensure_protocol_compliance
+    def set_parameters(
+        self: MrMtlPersonalizedProtocol, parameters: NDArrays, config: Config, fitting_round: bool
+    ) -> None:
+        """
+        The parameters being passed are to be routed to the initial global model to be used in a penalty term in
+        training the local model. Despite the usual FL setup, we actually never pass the aggregated model to the
+        **LOCAL** model. Instead, we use the aggregated model to form the MR-MTL penalty term.
+
+        NOTE; In MR-MTL, unlike Ditto, the local model weights are not synced across clients to the initial global
```
Comment (emersodb) [inline fl4health/mixins/personalized/mr_mtl.py:136]: Perhaps colon instead of semicolon? I imagine I made this typo in the MR-MTL client file as well. So if you could fix that would be great 🙂

File: tests/mixins/personalized/test_mr_mtl.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+        return self.train_loader, self.val_loader
+
+    def get_optimizer(self, config: Config) -> Optimizer | dict[str, Optimizer]:
+        return self.optimizers["local"]
+
+    def get_criterion(self, config: Config) -> _Loss:
+        return torch.nn.CrossEntropyLoss()
+
+
+class _TestDittoedClient(MrMtlPersonalizedMixin, _TestFlexibleClient):
```
Comment (emersodb) [inline tests/mixins/personalized/test_mr_mtl.py:39]: Should this be "Dittoed"?

File: tests/mixins/personalized/test_mr_mtl.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+from fl4health.mixins.core_protocols import FlexibleClientProtocol
+from fl4health.mixins.personalized import (
+    MrMtlPersonalizedMixin,
+    MrMtlPersonalizedProtocol,
+)
+from fl4health.parameter_exchange.packing_exchanger import FullParameterExchangerWithPacking
+from fl4health.parameter_exchange.parameter_packer import ParameterPackerAdaptiveConstraint
+
+
+class _TestFlexibleClient(FlexibleClient):
```
Comment (emersodb) [inline tests/mixins/personalized/test_mr_mtl.py:25]: Correct me if I'm wrong, but I think we have one of these classes for each of `test_mr_mtl.py`, `test_ditto.py`, and `test_adaptive_drift_constrained.py`. Could we consolidate them somewhere to reduce the duplicate code? We have a `tests/test_utils` for a few of these things.

File: tests/mixins/personalized/test_mr_mtl.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+
+    def get_criterion(self, config: Config) -> _Loss:
+        return torch.nn.CrossEntropyLoss()
+
+
+class _TestDittoedClient(MrMtlPersonalizedMixin, _TestFlexibleClient):
+    pass
+
+
+class _DummyParent:
```
Comment (emersodb) [inline tests/mixins/personalized/test_mr_mtl.py:43]: Same thing here, I believe there is one of these guys for Ditto tests as well?

File: fl4health/mixins/personalized/mr_mtl.py (hunk 1 above)
Comment (nerdai) [reply fl4health/mixins/personalized/mr_mtl.py:5]: lol I never even knew about WARNING but that's the main enum. I guess they have this alias for lazy people like me.

File: fl4health/mixins/personalized/mr_mtl.py (hunk 3 above)
Comment (nerdai) [reply fl4health/mixins/personalized/mr_mtl.py:91]: Yeah, sorry about the copy-and-paste issues with Ditto mixin... I used your suggested summary statement.

Comment (nerdai) [reply fl4health/mixins/personalized/mr_mtl.py:91]: I think the name is technically correct though right, since we're getting the model and then using it to set the `self.initial_global_model`. Like this method returns a model, so it is in that sense a getter.

But, to your point, we don't need to make a copy of the global model here do we? Like could I just make use of `self.get_model(config)`

File: fl4health/mixins/personalized/mr_mtl.py (hunk 4 above)
Comment (nerdai) [reply fl4health/mixins/personalized/mr_mtl.py:136]: changed in both

File: tests/mixins/personalized/test_mr_mtl.py (hunk 5 above)
Comment (nerdai) [reply tests/mixins/personalized/test_mr_mtl.py:39]: Thanks. Changed this to something more appropriate.

File: tests/mixins/personalized/test_mr_mtl.py (hunk 7 above)
Comment (nerdai) [reply tests/mixins/personalized/test_mr_mtl.py:43]: done!

File: fl4health/mixins/personalized/mr_mtl.py (hunk 2 above)
Comment (nerdai) [reply fl4health/mixins/personalized/mr_mtl.py:59]: I vaguely remember this now. Will run some tests on this in the AM. I think this may just be a static type error that was causing the issue.

This would get encountered if there is a parent in the MRO that doesn't accept args, kwargs. It's probably not something though that we should encounter given these mixins are meant to be mixed with `FlexibleClient` only. So, I'd be okay with just raising the error to indicate something off to the developer.

File: tests/mixins/personalized/test_mr_mtl.py (hunk 6 above)
Comment (nerdai) [reply tests/mixins/personalized/test_mr_mtl.py]: Note: in this instance with import statement style `from .conftest import ...` I actually like this here for two reasons:

1. its not really part of a library that we're importing from here
2. since multiple conftests can be defined and are automatically resolved with their hierarchical structure this indicates directly that its importing from the `conftest.py` at the same level as the test script

File: fl4health/mixins/personalized/mr_mtl.py (hunk 3 above)
Comment (emersodb) [reply fl4health/mixins/personalized/mr_mtl.py]: If you're good with the name we can stick with it. The `deep_copy` that's being done in the MR-MTL client was strictly for "safety" just in case somehow the `get_model` implemented by the user actually returned an already existing object. We want the global and local models to be guaranteed separate objects. So I believe we implemented the deep copy to ensure that was the case regardless.

Comment (nerdai) [reply fl4health/mixins/personalized/mr_mtl.py]: Oh, I didn't see us using `deep_copy` in `MrMtlClilent`. This part came from my copy over from `DittoPersonalizedMixin` to kickstart the mixin/protocol approach for MR-MTL. We used this `deep_copy` there for Nnunet stuff, I believe.

In `MrMtlClient`, we call `self.get_model(config)` under the `setup_client()` method:

```
    def setup_client(self, config: Config) -> None:
        """
        Set dataloaders, optimizers, parameter exchangers and other attributes derived from these.
        Then set initialized attribute to True.

        Args:
            config (Config): The config from the server.
        """
        # Need to setup the init global model here as well. It should be the same architecture as the model so
        # we reuse the get_model call. We explicitly send the model to the desired device. This is idempotent.
        self.initial_global_model = self.get_model(config).to(self.device)
        ...
```

(I think in our library the `get_model()` gets implemented as a method that returns a single model i.e., in a single memory location, and other times as a method that produces a new model copy/instance i.e., in a new memory location)

**Review by emersodb — APPROVED** [review]
